bip-csfs.mediawiki

```diff
@@ -0,0 +1,124 @@
+<pre>
+  BIP: TBD
+  Layer: Consensus (soft fork)
+  Title: CHECKSIGFROMSTACK
+  Author: Brandon Black <freedom@reardencode.com>
+  Comments-Summary: No comments yet.
+  Status: Draft
+  Type: Standards Track
+  Created: 2023-12-22
+  License: PD
+</pre>
+
+==Abstract==
+
+This BIP describes two new opcode for the purpose of checking cryptographic
+signatures in bitcoin scripts against data other than bitcoin transactions.
+
+==Summary==
+
+We propose replacing OP_NOP5 (0xb4) in bitcoin script with
+<code>OP_CHECKSIGFROMSTACKVERIFY</code>. When verifying taproot script spends having
+leaf version 0xc0 (as defined in BIP342), we propose <code>OP_CHECKSIGFROMSTACK</code>
+to replace <code>OP_SUCCESS204</code> (0xcc).
+
+<code>OP_CHECKSIGFROMSTACK</code> and <code>OP_CHECKSIGFROMSTACKVERIFY</code>
+have semantics similar to <code>OP_CHECKSIG</code> and
+<code>OP_CHECKSIGVERIFY</code> respectively, as specified below.
+
+For legacy and segwit v0 scripts, 32-byte and 33-byte keys are supported for
+BIP340 signature verification and ECDSA signature verification respectively.
+
+==Specification==
+
+* If fewer than 3 elements are on the stack, the script MUST fail and terminate immediately.
+* The public key (top element), message (second to top element), and signature (third from top element) are read from the stack.
+* For <code>OP_CHECKSIGFROMSTACK</code> the top three elements are popped from the stack.
+* If the public key size is zero, the script MUST fail and terminate immediately.
+* If the public key size is 32 bytes, it is considered to be a public key as described in BIP340:
+** If the signature is not the empty vector, the signature is validated against the public key and message according to BIP340. Validation failure in this case immediately terminates script execution with failure.
+* For legacy and segwit v0 scripts, if the public key size is 33 bytes and its first byte is 0x02 or 0x03, it is considered a compressed public key as described in BIP137:
+** If the signature is not the empty vector, the signature is validated against the public key and message using ECDSA. Validation failure in this case immediately terminates script execution with failure.
+* If the public key size is not zero, and it is not a BIP340 public key, nor a BIP137 compressed public key; the public key is of an unknown public key type, and no actual signature verification is applied. During script execution of signature opcodes they behave exactly as known public key types except that signature validation is considered to be successful.
```

Comment:
Public key types were only introduced in segwit v1 ([BIP 342](https://github.com/bitcoin/bips/blob/master/bip-0342.mediawiki#cite_note-6)), so this wouldn't apply to a version of this op code that also exists under the segwit v0 paradigm.

Reply:
Signature verification for CHECKSIGFROMSTACK is done the same regardless of the script context.